AppSyncRealTimeClient/ConnectionProvider/AppsyncRealtimeConnection/RealtimeConnectionProvider.swift

```diff
@@ -57,30 +59,43 @@ public class RealtimeConnectionProvider: ConnectionProvider {
     }
 
     public func write(_ message: AppSyncMessage) {
-        let signedMessage = interceptMessage(message, for: url)
-        let jsonEncoder = JSONEncoder()
-        do {
-            let jsonData = try jsonEncoder.encode(signedMessage)
-            guard let jsonString = String(data: jsonData, encoding: .utf8) else {
-                updateCallback(event: .error(ConnectionProviderError.jsonParse(message.id, nil)))
+
+        serialWriteQueue.async { [weak self] in
+            guard let self = self else {
                 return
             }
-            websocket.write(message: jsonString)
-        } catch {
-            AppSyncLogger.error(error)
-            switch message.messageType {
-            case .connectionInit:
-                serialConnectionQueue.async {[weak self] in
-                    guard let self = self else {
-                        return
-                    }
-                    self.status = .notConnected
-                    self.updateCallback(event: .error(ConnectionProviderError.connection))
+
+            let signedMessage = self.interceptMessage(message, for: self.url)
+            let jsonEncoder = JSONEncoder()
+            do {
+                let jsonData = try jsonEncoder.encode(signedMessage)
+                guard let jsonString = String(data: jsonData, encoding: .utf8) else {
+                    let jsonError = ConnectionProviderError.jsonParse(message.id, nil)
+                    self.updateCallback(event: .error(jsonError))
+                    return
+                }
+                self.websocket.write(message: jsonString)
+            } catch {
+                AppSyncLogger.error(error)
+                switch message.messageType {
+                case .connectionInit:
+                    self.receivedConnectionInit()
+                default:
+                    self.updateCallback(event: .error(ConnectionProviderError.jsonParse(message.id, error)))
                 }
-            default:
-                updateCallback(event: .error(ConnectionProviderError.jsonParse(message.id, error)))
             }
         }
+
+    }
+
+    func receivedConnectionInit() {
```

Comment:
nit- prefer to move it to the bottom with the other internal methods

Reply:
Updated

> Wasn't there another problem in `StarscreamAdapter` where the websocket was sending messages to the delegate on the the main thread?
> 
> In Websocket.swift
> 
> ```
> // Where the callback is executed. It defaults to the main UI thread queue.
>     public var callbackQueue = DispatchQueue.main
> ```

Added a callback queue as well. This should work without the new callback but adding now to be safe.